aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -147,23 +181,40 @@ void TensorIterator::compute_types() {
           op.dtype = common_dtype;
         }
       }
-    }
-  }
 
-  for (auto& op : operands_) {
-    auto& tensor = op.tensor;
-    if (!tensor.defined()) {
-      continue;
-    }
-    if (op.device != tensor.device() || op.dtype != tensor.scalar_type()) {
-      if (op.is_output) {
-        AT_ERROR("output with device ", tensor.device(), " and dtype ", tensor.scalar_type(),
-                 " doesn't match the desired device ", op.device, " and dtype ", op.dtype);
-      } else if (tensor.dim() == 0) {
-        tensor = tensor.to(op.options());
-      } else {
-        AT_ERROR("expected device ", op.device, " and dtype ", op.dtype,
-                 " but got device ", tensor.device(), " and dtype ", tensor.scalar_type());
+      if (op.tensor.defined() && op.tensor.scalar_type() != common_dtype)
+      {
+        if (op.is_output) {
+          if (!can_promote(common_dtype, op.tensor.scalar_type())) {
```

Comment:
I don't see how casting of outputs is handled here. What happens if the common_dtype doesn't match the output type?

Reply:
This is still converting the operands to the result type, not the result of the operation as in numpy. I'm not sure if this is sufficient.

If an output tensor with a dtype is specified, it's included in the result_type calculation... e.g. if you were to do torch.add(float_tensor, float_tensor, out=double_tensor), ... the common_dtype is going to be derived from `operands_ = [float_tensor, double_tensor]`, and so each float_tensor operand gets converted to double and no final conversion is necessary. Is there a case where the input operands can all be promoted to the output type but where the output type would not be the common dtype?

Reply:
ah, right, with same_kind casting double->float requires a cast.